### ISSUE TITLE
`TypeError: Cannot perform Construct on a detached ArrayBuffer` on Deno

### ISSUE THREAD

**Issue by ariofrio** [httptoolkit/xz-decompress#1]
This seems to be caused by not updating the reference to the underlying WASM memory buffer before calling `Uint8Array.subarray`. I have not tested the library in Node, but if it works there it might because internal implementation details of Node do not invalidate (e.g. grow) the WASM memory buffer before `supplyInput` is called, but the Deno allocator does.

```
error: Uncaught (in promise) TypeError: Cannot perform Construct on a detached ArrayBuffer
    at Uint8Array.subarray (<anonymous>)
    at XzContext.supplyInput (file:///~/Library/Caches/deno/npm/registry.npmjs.org/xz-decompress/0.2.1/dist/package/xz-decompress.js:149:38)
    at Object.pull (file:///~/Library/Caches/deno/npm/registry.npmjs.org/xz-decompress/0.2.1/dist/package/xz-decompress.js:240:27)
    at eventLoopTick (ext:core/01_core.js:183:11)
```

The solution is to add a call to `this._refresh()` to the beginning of `XzContext.supplyInput()`:

```
        supplyInput(sourceDataUint8Array) {
          this._refresh(); // <===
          const inBuffer = this.mem8.subarray(this.inStart, this.inEnd);
          inBuffer.set(sourceDataUint8Array, 0);
          this.exports.supply_input(this.ptr, sourceDataUint8Array.byteLength);
          this._refresh();
        }
```

**Comment by pimterry** [comment]
Sorry, I have no idea about that :smile:. It sounds like this is a bug that should be reported to either Node or Deno though to be honest.

I'm not an expert on the relevant WASM internals, but personally I'd assume this is Deno failing incorrectly, rather than Node, because (as far as I'm aware!) this library does work fine in web browsers, who I expect to be more likely to accurately represent the spec than either standalone VM.

In general, Deno isn't officially supported (I've never tried this with Deno myself, there's no Deno test suite here, and since I'm not using it I don't have any time to add that). I'm open to fixes to improve that as required, if somebody else wants to make that happen, especially if we can set up a Deno test suite here to help confirm those. I would want clear confirmation that there's a good reason for the differences first though, rather than just adding a workaround for a Deno bug.

**Comment by pimterry** [comment]
I've now had time to hunt this down (seems to be reliably reproducible when doing two decompressions in parallel) and properly test your suggested fix here - which seems to work nicely. That's just been released as v0.2.2. Thanks!
